peer.go

```diff
@@ -1,516 +1,18 @@
-	p.StatsMtx.Lock()
-
-	// Limit to one version message per peer.
-	if p.versionKnown {
-		p.logError("Only one version message per peer is allowed %s.",
-			p)
-		p.StatsMtx.Unlock()
-
-		// Send an reject message indicating the version message was
-		// incorrectly sent twice and wait for the message to be sent
-		// before disconnecting.
-		p.PushRejectMsg(msg.Command(), wire.RejectDuplicate,
-			"duplicate version message", nil, true)
-
-		p.Disconnect()
-		return
-	}
-
-	// Negotiate the protocol version.
-	p.protocolVersion = minUint32(p.protocolVersion, uint32(msg.ProtocolVersion))
-	p.versionKnown = true
-	peerLog.Debugf("Negotiated protocol version %d for peer %s",
-		p.protocolVersion, p)
-	p.lastBlock = msg.LastBlock
-	p.startingHeight = msg.LastBlock
-
-	// Set the supported services for the peer to what the remote peer
-	// advertised.
-	p.services = msg.Services
-
-	// Set the remote peer's user agent.
-	p.userAgent = msg.UserAgent
-
-	// Set the peer's time offset.
-	p.timeOffset = msg.Timestamp.Unix() - time.Now().Unix()
-
-	// Set the peer's ID.
-	p.id = atomic.AddInt32(&nodeCount, 1)
-
-	p.StatsMtx.Unlock()
-
-	// Choose whether or not to relay transactions before a filter command
-	// is received.
-	p.relayMtx.Lock()
-	p.disableRelayTx = msg.DisableRelayTx
-	p.relayMtx.Unlock()
-
-	// Inbound connections.
-	if p.inbound {
-		// Set up a NetAddress for the peer to be used with AddrManager.
-		// We only do this inbound because outbound set this up
-		// at connection time and no point recomputing.
-		na, err := newNetAddress(p.conn.RemoteAddr(), p.services)
-		if err != nil {
-			p.logError("Can't get remote address: %v", err)
-			p.Disconnect()
-			return
-		}
-		p.na = na
-
-		// Send version.
-		err = p.pushVersionMsg()
-		if err != nil {
-			p.logError("Can't send version message to %s: %v",
-				p, err)
-			p.Disconnect()
-			return
-		}
-	}
-
-	// Send verack.
-	p.QueueMessage(wire.NewMsgVerAck(), nil)
-
-	// Update the address manager and request known addresses from the
-	// remote peer for outbound connections.  This is skipped when running
-	// on the simulation test network since it is only intended to connect
-	// to specified peers and actively avoids advertising and connecting to
-	// discovered peers.
-	if !cfg.SimNet {
-		p.updateAddresses(msg)
-	}
-
-	// Add the remote peer time as a sample for creating an offset against
-	// the local clock to keep the network time in sync.
-	p.server.timeSource.AddTimeSample(p.addr, msg.Timestamp)
-
-	// Signal the block manager this peer is a new sync candidate.
-	p.server.blockManager.NewPeer(p)
-
-	// TODO: Relay alerts.
-}
-
 // pushTxMsg sends a tx message for the provided transaction hash to the
 // connected peer.  An error is returned if the transaction hash is not known.
-func (p *peer) pushTxMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
+func pushTxMsg(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```

Comment:
This should be moved onto the server type in `server.go` imo.  It's only used by its `handleGetDataMsg` and seems to more logically fit there.  Plus, by moving this along with `pushBlockMsg` and `pushMerkleBlockMsg`, the entire `peer.go` file can be removed.

```diff
diff --git a/server.go b/server.go
index 4072f8d..03adecb 100644
--- a/server.go
+++ b/server.go
@@ -288,6 +288,33 @@ func (s *server) handleHeadersMsg(p *peer.Peer, msg *wire.MsgHeaders) {
        s.blockManager.QueueHeaders(msg, p)
 }

+// pushTxMsg sends a tx message for the provided transaction hash to the
+// connected peer.  An error is returned if the transaction hash is not known.
+func (s *server) pushTxMsg(p *peer.Peer, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
+       ...
+}
+
 // handleGetData is invoked when a peer receives a getdata bitcoin message
 // and is used to deliver block and transaction information.
 func (s *server) handleGetDataMsg(p *peer.Peer, msg *wire.MsgGetData) {
@@ -313,7 +340,7 @@ func (s *server) handleGetDataMsg(p *peer.Peer, msg *wire.MsgGetData) {
```